src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,656 @@
+                    {
+                        UpdateExpressionInOriginalFunction(semanticDocument, expressionCopy, oldMethodDeclaration, parameterName, editor, allOccurrences, cancellationToken);
+
+                        var parameterType = semanticDocument.SemanticModel.GetTypeInfo(expressionCopy, cancellationToken).Type ?? semanticDocument.SemanticModel.Compilation.ObjectType;
+                        var refKind = syntaxFacts.GetRefKindOfArgument(expressionCopy);
+                        var parameter = generator.ParameterDeclaration(name: parameterName, type: generator.TypeExpression(parameterType), refKind: refKind);
+                        editor.InsertParameter(oldMethodDeclaration, insertionIndex, parameter);
+                    }
+
+                    var newRoot = editor.GetChangedRoot();
+                    modifiedSolution = modifiedSolution.WithDocumentSyntaxRoot(firstCallSite.Id, newRoot);
+                }
+            }
+
+            return modifiedSolution;
+        }
+
+        /// <summary>
+        /// Ties the identifiers within the expression back to their associated parameter
+        /// </summary>
+        public static Dictionary<TIdentifierNameSyntax, IParameterSymbol> MapExpressionToParameters(SemanticDocument document, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            var nameToParameterDict = new Dictionary<TIdentifierNameSyntax, IParameterSymbol>();
+            var variablesInExpression = expression.DescendantNodes().OfType<TIdentifierNameSyntax>();
+            var semanticModel = document.SemanticModel;
+
+            foreach (var variable in variablesInExpression)
+            {
+                var symbol = semanticModel.GetSymbolInfo(variable, cancellationToken).Symbol;
+                if (symbol != null && symbol is IParameterSymbol parameterSymbol)
+                {
+                    if (!nameToParameterDict.ContainsKey(variable))
+                    {
+                        nameToParameterDict.Add(variable, parameterSymbol);
+                    }
+                }
+            }
+            return nameToParameterDict;
+        }
+
+        /// <summary>
+        /// Determines if the expression is contained within something that is "parameterized"
+        /// </summary>
+        private (bool isParameterized, bool hasOptionalParameter) ExpressionWithinParameterizedMethod(SemanticDocument semanticDocument,
+                                                                                                      TExpressionSyntax expression,
+                                                                                                      CancellationToken cancellationToken)
+        {
+            var methodSymbol = GetMethodSymbolFromExpression(semanticDocument, expression, cancellationToken);
+
+            var variablesInExpression = expression.DescendantNodes().OfType<TIdentifierNameSyntax>();
```

Comment:
i don't understand why we need to check descendant nodes to "Determines if the expression is contained within something that is "parameterized""

Reply:
I need to retrieve the identifiers in the expression to see if they map back to the parameters, how would I do it otherwise?

Reply:
i guess i don't understand what this method is trying to do :)

Reply:
It is looking at the identifiers within the expression and seeing that they all map back to parameters because, if not, then the code actions should not show up.

Reply:
ok.  i'm saying: the method's name and decsription are not clear.  It is askign if the expression is inside a method.  but then it does a lot of extra stuff.  either that stuff is unnecessary.  or the name/docs are wrong.

Reply:
I'll change the name to "ShouldExpressionDisplayCodeAction"?